### PR TITLE
Use solang's heap implementation on Solana

### PR DESCRIPTION
**Pull request by seanyoung**
As of Solana v1.11, `sol_alloc_free_` system call is no longer available. This means that programs are expected to provide their own heap impl.

Also enable `push()` and `pop()` for memory arrays on Solana.

### CODE REVIEW

File: tests/solana.rs [hunk 1]
```diff
@@ -854,7 +935,10 @@ impl<'a> SyscallObject<UserError> for SyscallLogData<'a> {
         memory_mapping: &mut MemoryMapping,
         result: &mut Result<u64, EbpfError<UserError>>,
     ) {
+        self.context.heap_verify();
+
         if let Ok(mut vm) = self.context.vm.try_borrow_mut() {
+            println!("sol_log_data");
```
Comment (LucasSte) [inline tests/solana.rs:941]: Is this for debugging?

File: tests/solana_tests/arrays.rs [hunk 2]
```diff
@@ -1024,3 +1024,258 @@ fn initialization_with_literal() {
...
+            function test() public {
+                int[] bar = new int[](1);
+                bar[0] = 12;
+                bar.pop();
+
+                assert(bar[0] == 12);
+            }
+        }"#,
+    );
+
```
Comment (LucasSte) [inline tests/solana_tests/arrays.rs:1278]: There are no tests for pushes and pops in the same function.

Comment (LucasSte) [reply tests/solana_tests/arrays.rs:1278]: Pushing and poping several times tests your heap implementation.

**Review by LucasSte — CHANGES_REQUESTED** [review]

File: tests/solana.rs (hunk 1 above)
Comment (seanyoung) [reply tests/solana.rs:941]: For testing purposes it is useful to know that `sol_log_data` is called. The mock solana (and substrate) print lots of stuff which helps debugging.

File: src/emit/mod.rs [hunk 3]
```diff
@@ -5062,10 +5077,10 @@ pub trait TargetRuntime<'a> {
                         bin.builder.build_int_add(len, len, "hex_len")
                     }
                     Type::Uint(bits) if *spec == FormatArg::Hex => {
-                        bin.context.i32_type().const_int(bits as u64 / 8 + 2, false)
+                        bin.context.i32_type().const_int(bits as u64 / 4 + 2, false)
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply src/emit/mod.rs:5080]: Because it causes a buffer overrun. A converting an unsigned int to a hex string with the `0x` prefix can be a maximum of `bits / 4 + 2`, characters, not `bits / 8 + 2`.

File: tests/solana.rs (hunk 1 above)
Comment (seanyoung) [reply tests/solana.rs]: Now it also logs the argument to `sol_log_data`

File: tests/solana_tests/arrays.rs (hunk 2 above)
Comment (seanyoung) [reply tests/solana_tests/arrays.rs:1278]: Good point. I've added a test for this.

File: src/emit/ethabiencoder.rs [hunk 4]
```diff
@@ -654,40 +658,46 @@ impl<'a, 'b> EncoderBuilder<'a, 'b> {
...
+                        )
+                        .into(),
+                ],
+                "",
+            );
 
-        let mut output = output;
-        let mut offset = self.offset;
-        let mut dynamic = unsafe { binary.builder.build_gep(output, &[self.offset], "") };
+            let mut output = output;
```
Comment (LucasSte) [inline src/emit/ethabiencoder.rs:682]: You can declare the function argument as mutable to avoid creating another variable with the same name.

File: tests/solana_tests/arrays.rs [hunk 5]
```diff
@@ -1024,3 +1024,298 @@ fn initialization_with_literal() {
...
+        r#"
+        contract foo {
+            function test() public {
+                uint32[] bar1 = new uint32[](0);
+                uint32[] bar2 = new uint32[](0);
+
+                // each time we call a system call, the heap is checked
+                // for consistency. So do a print() after each operation
+                for (uint64 i = 1; i < 160; i++) {
+                    if ((i % 10) == 0) {
```
Comment (LucasSte) [inline tests/solana_tests/arrays.rs:1295]: Test a vector of strings too (we have had problems with these before).

**Review by LucasSte — CHANGES_REQUESTED** [review]

File: tests/solana_tests/yul.rs [hunk 6]
```diff
@@ -79,7 +79,7 @@ contract testing  {
     assert_eq!(
         returns,
         vec![
-            Token::Uint(U256::from(12884901888u64)),
+            Token::Uint(U256::from(12884901920u64)),
```
[the remark this answers is not in the document]
Comment (seanyoung) [reply tests/solana_tests/yul.rs]: agreed

File: CHANGELOG.md [hunk 7]
```diff
@@ -2,6 +2,13 @@
 All notable changes to [Solang](https://github.com/hyperledger/solang/)
 will be documented here.
 
+## Unreleased
+
+### Added
+- On Solana, programs now use a custom heap implementation, just like on
+  Substrate. As result it is now possible to `.push()` and `.pop()` on
```
Comment (LucasSte) [inline CHANGELOG.md:9]:
```suggestion
  Substrate. As result, it is now possible to `.push()` and `.pop()` on
```
A comma is missing.

File: src/emit/ethabiencoder.rs (hunk 4 above)
Comment (LucasSte) [reply src/emit/ethabiencoder.rs:681]: @seanyoung You declared the argument as mutable, but you are still creating the variable here!

File: stdlib/heap.c [hunk 8]
```diff
@@ -118,7 +142,7 @@ void *__realloc(void *m, size_t size)
     else
     {
         void *n = __malloc(size);
-        __memcpy8(n, m, size / 8);
+        __memcpy8(n, m, (size + 7) / 8);
```
Comment (LucasSte) [inline stdlib/heap.c:145]: Why is this necessary? This was not in the previous version of your PR.

**Review by LucasSte — CHANGES_REQUESTED** [review]

File: stdlib/heap.c (hunk 8 above)
Comment (seanyoung) [reply stdlib/heap.c:145]: Another bug fix in the heap implementation

Comment (seanyoung) [reply stdlib/heap.c]: Changed again!

**Comment by seanyoung** [comment]
Fixes #991

File: stdlib/heap.c [hunk 9]
```diff
@@ -109,16 +132,21 @@ void *__realloc(void *m, size_t size)
...
     }
     else
     {
+        uint32_t len = cur->length;
+        if (size < len)
+            len = size;
+
         void *n = __malloc(size);
-        __memcpy8(n, m, size / 8);
+        __memcpy8(n, m, (len + 7) / 8);
```
Comment (LucasSte) [inline stdlib/heap.c:149]: This deserves a comment

**Review by LucasSte — CHANGES_REQUESTED** [review]

File: src/emit/mod.rs [hunk 10]
```diff
@@ -5075,10 +5090,10 @@ pub trait TargetRuntime<'a> {
                     }
                     // bits / 3 is a rough over-estimate of how many decimals we need
```
Comment (LucasSte) [inline src/emit/mod.rs:5091]: You've updated the division, but this comment also needs updating.

**Review by LucasSte — APPROVED** [review]
Please, fix the comment before merging.
Also, update Solana's version to v1.11 in our integration tests to make sure this works. This can be done in another PR.

File: src/emit/mod.rs (hunk 10 above)
Comment (seanyoung) [reply src/emit/mod.rs]: Fixed

**Comment by seanyoung** [comment]
> Please, fix the comment before merging. Also, update Solana's version to v1.11 in our integration tests to make sure this works. This can be done in another PR.

The v1.11 docker image is broken: 
```bash
$ podman run --rm -it solanalabs/solana:v1.11.10
...
[2022-09-05T13:01:44.635456291Z ERROR solana_ledger::blockstore] Unable to increase the maximum open file descriptor limit to 1000000 from 524288
```
I don't know how to fix this.
